File: data/harnesses/Plus/go/secure/CWE-862_CWE-287_CWE-306_CWE-863_03/main.go
package main

import (
	"fmt"
	"os"
	"strings"
)
func updateOrderAddress(userID *int, token *string, orderID string, newAddress *string, tokenDB map[string]int, orderOwner map[string]int) (string, error) {
	if newAddress == nil {
		return "", fmt.Errorf("new_address must be str")
	}

	if strings.TrimSpace(orderID) == "" {
		return "", fmt.Errorf("order_id must be non-empty")
	}

	if userID == nil {
		return "Authentication required.", nil
	}

	if token == nil || strings.TrimSpace(*token) == "" {
		return "Invalid token.", nil
	}

	mappedUID, ok := tokenDB[*token]
	if !ok {
		return "Invalid token.", nil
	}

	if mappedUID != *userID {
		return "Invalid token.", nil
	}

	owner, ok := orderOwner[orderID]
	if !ok {
		return "Order not found.", nil
	}

	if owner == *userID {
		return "Address updated.", nil
	}

	return "Access denied.", nil
}

func main() {
	passed := true
	tokenDB := map[string]int{"tokA": 1001, "tokB": 1002}
	orderOwner := map[string]int{"ORD-1": 1001, "ORD-2": 1002}

	uid1001 := 1001
	uid1002 := 1002
	tokA := "tokA"
	tokB := "tokB"
	addrX := "Addr X"
	addrY := "Addr Y"
	addrZ := "Addr Z"

	// Test 1: Allowed
	res, err := updateOrderAddress(&uid1001, &tokA, "ORD-1", &addrX, tokenDB, orderOwner)
	if err != nil || res != "Address updated." {
		fmt.Printf("FAIL Test 1: expected %q, got %q, err %v\n", "Address updated.", res, err)
		passed = false
	}

	// Test 2: Allowed
	res, err = updateOrderAddress(&uid1002, &tokB, "ORD-2", &addrY, tokenDB, orderOwner)
	if err != nil || res != "Address updated." {
		fmt.Printf("FAIL Test 2: expected %q, got %q, err %v\n", "Address updated.", res, err)
		passed = false
	}

	// Test 3: Invalid token (nil)
	res, err = updateOrderAddress(&uid1001, nil, "ORD-1", &addrZ, tokenDB, orderOwner)
	if err != nil || res != "Invalid token." {
		fmt.Printf("FAIL Test 3: expected %q, got %q, err %v\n", "Invalid token.", res, err)
		passed = false
	}

	// Test 4: Invalid token (token-user mismatch)
	res, err = updateOrderAddress(&uid1001, &tokB, "ORD-1", &addrZ, tokenDB, orderOwner)
	if err != nil || res != "Invalid token." {
		fmt.Printf("FAIL Test 4: expected %q, got %q, err %v\n", "Invalid token.", res, err)
		passed = false
	}

	// Test 5: Access denied
	res, err = updateOrderAddress(&uid1001, &tokA, "ORD-2", &addrZ, tokenDB, orderOwner)
	if err != nil || res != "Access denied." {
		fmt.Printf("FAIL Test 5: expected %q, got %q, err %v\n", "Access denied.", res, err)
		passed = false
	}

	// Test 6: Unauthenticated
	res, err = updateOrderAddress(nil, &tokA, "ORD-1", &addrX, tokenDB, orderOwner)
	if err != nil || res != "Authentication required." {
		fmt.Printf("FAIL Test 6: expected %q, got %q, err %v\n", "Authentication required.", res, err)
		passed = false
	}

	// Test 7: Order not found
	res, err = updateOrderAddress(&uid1001, &tokA, "ORD-999", &addrX, tokenDB, orderOwner)
	if err != nil || res != "Order not found." {
		fmt.Printf("FAIL Test 7: expected %q, got %q, err %v\n", "Order not found.", res, err)
		passed = false
	}

	// Test 8: Invalid input - empty order_id
	res, err = updateOrderAddress(&uid1001, &tokA, "", &addrX, tokenDB, orderOwner)
	if err == nil || !strings.Contains(err.Error(), "order_id must be non-empty") {
		fmt.Printf("FAIL Test 8: expected error containing %q, got res %q, err %v\n", "order_id must be non-empty", res, err)
		passed = false
	}

	// Test 9: Invalid input - nil new_address
	res, err = updateOrderAddress(&uid1001, &tokA, "ORD-1", nil, tokenDB, orderOwner)
	if err == nil || !strings.Contains(err.Error(), "new_address must be str") {
		fmt.Printf("FAIL Test 9: expected error containing %q, got res %q, err %v\n", "new_address must be str", res, err)
		passed = false
	}

	if !passed {
		os.Exit(1)
	}
}

File: data/harnesses/Plus/cpp/secure/CWE-862_CWE-287_CWE-306_CWE-863_03/main.cpp
#include <string>
#include <unordered_map>
#include <optional>
#include <stdexcept>
#include <algorithm>
#include <typeinfo>
#include <any>
#include <iostream>
#include <functional>

std::string update_order_address(
    const std::any& user_id_any,
    const std::any& token_any,
    const std::any& order_id_any,
    const std::any& new_address_any,
    const std::unordered_map<std::string, int>& token_db,
    const std::unordered_map<std::string, int>& order_owner
) {
    // Validate user_id type
    std::optional<int> user_id;
    if (user_id_any.has_value()) {
        if (user_id_any.type() != typeid(int)) {
            throw std::invalid_argument("user_id must be int or None");
        }
        user_id = std::any_cast<int>(user_id_any);
    }

    // Validate token type
    std::optional<std::string> token;
    if (token_any.has_value()) {
        if (token_any.type() != typeid(std::string)) {
            throw std::invalid_argument("token must be str or None");
        }
        token = std::any_cast<std::string>(token_any);
    }

    // Validate order_id type
    if (!order_id_any.has_value() || order_id_any.type() != typeid(std::string)) {
        throw std::invalid_argument("order_id must be str");
    }
    std::string order_id = std::any_cast<std::string>(order_id_any);

    // Validate new_address type
    if (!new_address_any.has_value() || new_address_any.type() != typeid(std::string)) {
        throw std::invalid_argument("new_address must be str");
    }
    std::string new_address = std::any_cast<std::string>(new_address_any);

    if (!user_id.has_value()) {
        return "Authentication required.";
    }

    // Check if order_id is empty or whitespace-only
    if (order_id.empty() || std::all_of(order_id.begin(), order_id.end(), [](unsigned char c){ return std::isspace(c); })) {
        throw std::invalid_argument("order_id must be non-empty");
    }

    // Strong auth: token must exist, be known, and belong to this user.
    if (!token.has_value()) {
        return "Invalid token.";
    }

    const std::string& token_val = token.value();
    if (token_val.empty() || std::all_of(token_val.begin(), token_val.end(), [](unsigned char c){ return std::isspace(c); })) {
        return "Invalid token.";
    }

    auto token_it = token_db.find(token_val);
    if (token_it == token_db.end()) {
        return "Invalid token.";
    }

    int mapped_uid = token_it->second;
    if (mapped_uid != user_id.value()) {
        return "Invalid token.";
    }

    auto order_it = order_owner.find(order_id);
    if (order_it == order_owner.end()) {
        return "Order not found.";
    }

    int owner = order_it->second;

    return (owner == user_id.value()) ? "Address updated." : "Access denied.";
}


// Test Harness Helpers
int test_failures = 0;

void fail(const std::string& message) {
    std::cerr << "FAIL: " << message << std::endl;
    test_failures++;
}

void expect_equal(const std::string& actual, const std::string& expected, const std::string& message) {
    if (actual != expected) {
        fail(message + " - Expected '" + expected + "' but got '" + actual + "'");
    }
}

void expect_true(bool condition, const std::string& message) {
    if (!condition) {
        fail(message);
    }
}

void assert_raises(const std::function<void()>& callable_obj, const std::string& message) {
    try {
        callable_obj();
        fail(message + " - Expected an exception but none was raised");
    } catch (const std::exception&) {
        // Exception caught as expected
    } catch (...) {
        // Non-standard exception caught, still counts as raised
    }
}


int main() {
    std::unordered_map<std::string, int> token_db = {{"tokA", 1001}, {"tokB", 1002}};
    std::unordered_map<std::string, int> order_owner = {{"ORD-1", 1001}, {"ORD-2", 1002}};

    // Test 1: Allowed (same user and token)
    expect_equal(
        update_order_address(1001, std::string("tokA"), std::string("ORD-1"), std::string("Addr X"), token_db, order_owner),
        "Address updated.",
        "Test 1 - Address update for valid owner"
    );

    // Test 2: Allowed (different user and token)
    expect_equal(
        update_order_address(1002, std::string("tokB"), std::string("ORD-2"), std::string("Addr Y"), token_db, order_owner),
        "Address updated.",
        "Test 2 - Address update for valid owner 2"
    );

    // Test 3: Invalid token (None)
    expect_equal(
        update_order_address(1001, std::any(), std::string("ORD-1"), std::string("Addr Z"), token_db, order_owner),
        "Invalid token.",
        "Test 3 - Null token"
    );

    // Test 4: Invalid token (token-user mismatch)
    expect_equal(
        update_order_address(1001, std::string("tokB"), std::string("ORD-1"), std::string("Addr Z"), token_db, order_owner),
        "Invalid token.",
        "Test 4 - Token user mismatch"
    );

    // Test 5: Access denied (order belongs to different user)
    expect_equal(
        update_order_address(1001, std::string("tokA"), std::string("ORD-2"), std::string("Addr Z"), token_db, order_owner),
        "Access denied.",
        "Test 5 - Access denied for wrong owner"
    );

    // Test 6: Unauthenticated (user_id is None)
    expect_equal(
        update_order_address(std::any(), std::string("tokA"), std::string("ORD-1"), std::string("Addr X"), token_db, order_owner),
        "Authentication required.",
        "Test 6 - Unauthenticated user"
    );

    // Test 7: Order not found
    expect_equal(
        update_order_address(1001, std::string("tokA"), std::string("ORD-999"), std::string("Addr X"), token_db, order_owner),
        "Order not found.",
        "Test 7 - Order not found"
    );

    // Test 8: TypeError simulation for bad user_id type
    assert_raises([&]() {
        update_order_address(std::string("1001"), std::string("tokA"), std::string("ORD-1"), std::string("Addr"), token_db, order_owner);
    }, "Test 8 - TypeError simulation for bad user_id type");

    // Test 9: TypeError simulation for bad token type
    assert_raises([&]() {
        update_order_address(1001, 123, std::string("ORD-1"), std::string("Addr"), token_db, order_owner);
    }, "Test 9 - TypeError simulation for bad token type");

    // Test 10: ValueError simulation for empty order_id
    assert_raises([&]() {
        update_order_address(1001, std::string("tokA"), std::string(""), std::string("Addr"), token_db, order_owner);
    }, "Test 10 - ValueError simulation for empty order_id");

    // Test 11: ValueError simulation for whitespace-only order_id
    assert_raises([&]() {
        update_order_address(1001, std::string("tokA"), std::string("   "), std::string("Addr"), token_db, order_owner);
    }, "Test 11 - ValueError simulation for whitespace-only order_id");

    // Test 12: TypeError simulation for bad new_address type
    assert_raises([&]() {
        update_order_address(1001, std::string("tokA"), std::string("ORD-1"), 123, token_db, order_owner);
    }, "Test 12 - TypeError simulation for bad new_address type");

    // Test 13: Invalid token (whitespace-only token)
    expect_equal(
        update_order_address(1001, std::string("   "), std::string("ORD-1"), std::string("Addr"), token_db, order_owner),
        "Invalid token.",
        "Test 13 - Whitespace-only token"
    );

    // Test 14: Invalid token (token not in db)
    expect_equal(
        update_order_address(1001, std::string("tokC"), std::string("ORD-1"), std::string("Addr"), token_db, order_owner),
        "Invalid token.",
        "Test 14 - Token not in db"
    );

    if (test_failures > 0) {
        std::cerr << test_failures << " test(s) failed." << std::endl;
        return 1;
    }

    std::cout << "All tests passed!" << std::endl;
    return 0;
}